### ISSUE TITLE
async await generator triggers Chrome Web Store's Obfuscation violation

### ISSUE THREAD

**Issue by adamhwang** [microsoft/TypeScript#41242]
I have a [Chrome extension](https://github.com/adamhwang/ita-matrix-powertools) that was first published to the Chrome Web Store in Feb 2020. A Typescript transpiled script was first published on Sep 20 2020. Prior to that, the script was packaged with webpack. On Oct 24 2020, Google disabled the extension for violating its [Obfuscation](https://developer.chrome.com/webstore/troubleshooting#obfuscation) policy (Red Titanium).

While Google did not identify the offending code, the most obfuscated code appeared to be the `__generator` function generated for the async/await pollyfill. I cannot be sure that the webpack boilerplate did not trigger the removal however, but I have published a new version without the `__generator` functions and will keep this issue updated.

The immediate workaround was to target es2017 to avoid transpiling async/await unto generators. Ideally though, I'd prefer to avoid having to add babel while still being able to target es3 because this script is also distributed as a userscript.

**TypeScript Version:**  3.9.7

**Search Terms:** typescript async await __awaiter __generator minify uglify obfuscate google chrome web store red titanium

**Code**

```ts
async function hello () {
  console.log('world');
}
```

**Expected behavior:** Human readable polyfills and let downstream minifiers/uglifiers do their jobs

**Actual behavior:**
```ts
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
function hello() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            console.log('world');
            return [2 /*return*/];
        });
    });
}
 ```

**Playground Link:** [Provided](https://www.typescriptlang.org/play?target=0#code/IYZwngdgxgBAZgV2gFwJYHsIwBYFMA2+6MAFAJQwDeAUAJBSYjr64B0RA5iQOQDu6AJ3wATbmQDc1AL5A)

**Related Issues:** #14689

**Comment by kitsonk** [comment]
I believe these helpers are intentionally terse, for people who do not wish to use further build steps.

There are alternatives, like `--importHelpers` and `--noEmitHelpers` [compiler options](https://www.typescriptlang.org/docs/handbook/compiler-options.html) which would allow you (or your build tooling) to provide alternative implementations.

**Comment by DanielRosenwasser** [comment]
I don't really understand this issue and whether it's actionable. What would need to change for this code to be "acceptable"?

**Comment by adamhwang** [comment]
I mainly just wanted to track this as a use case and an issue for others that may stumble upon this, especially as I noticed that there was another related issue opened with the same ask but without a specific problem to address.

I'm honestly not sure how how this could be defined more objectively and less subjectively. At a minimum, I think replacing the single-character variable names may make it at least _look_ readable, but who knows what kind of tools and metrics are being used to measure against that requirement.

**Comment by DanielRosenwasser** [comment]
I guess the workaround I'd suggest is just compile once in ES2020 and once in ES3 to have multiple builds. It's unfortunate that Google has no obvious means to reach out to appeal for any sort of exemption (e.g. no "Contact Us" link on that page).

**Comment by DanielRosenwasser** [comment]
@evmar @mprobst @TimvdLippe do any of you know if there's someone who could help in this case? Seems like there's some set of compiled output that the Chrome Web Store may want to consider as an exception to the current obfuscation checks.

**Comment by TimvdLippe** [comment]
@dotproto is the best point-of-contact.

**Comment by dotproto** [comment]
Hey all, just wanted to pop in offer my two cents that seems like a mistake on our end. I'm going to reach out to the review after I post this to open a discussion about this issue and see if we can.

It has been a while since I last worked with typescript. Is it possible to set a compiler option to use JS's built in async support? One of my concerns is that Malicious actors may try to use something that looks like TS's awaiter and generator, but slips in a couple extra bits of code that allow for underhanded code execution (e.g. underhanded-c.org).

**Comment by adamhwang** [comment]
@dotproto @DanielRosenwasser Thank you. That's the workaround we went with. By targeting es2017, we can utilize the built in js async/await support which should be fine for the chrome and other browser extensions. The drawback though is that we were using the same script as a userscript and a bookmarklet, which will have some reduced browser support
